dag_in_context/src/optimizations/hackers_delight.egg

```diff
@@ -0,0 +1,90 @@
+;; Hacker's delight optimizations
+
+(ruleset hacker)
+
+;; A simple analysis to identify loops that run exactly #popcount times
+
+;; IsIsEven e x => e is a boolean expression that checks whether x is an even number
+(relation IsIsEven (Expr Expr))
+
+(rule (
+    (= two (Const (Int 2) ty ctx))
+    (= e (Bop (Eq) x (Bop (Mul) (Bop (Div) x two) two)))
+) (
+    (IsIsEven e x)
+) :ruleset hacker)
+
+;; NTZIterations lp n pos => loop lp runs exactly number_of_trailing_zeros(n) times at index pos
+(relation NTZIterations (Expr Expr i64))
+
+(rule (
+    ;; Grab the outer if
+    (= outerif (If cond inputs evenbr oddbr))
+    ;; There exists an argument n
+    (= n (Get inputs i))
+    ;; The condition is on the parity of n
+    (IsIsEven cond n)
+    ;; In the even/true branch, there is a loop
+    (= evenbr (DoWhile lp_inputs lp_pred_outputs))
+    ;; n is passed into to the loop
+    (= (Get lp_inputs j) (Get (Arg _ty1 _ctx1) i))
+    ;; the loop continues as long as n / 2 is even
+    (= two (Const (Int 2) _ty2 _ctx2))
+    (= nd2 (Bop (Div) (Get (Arg _ty3 _ctx3) j) two))
+    (IsIsEven (Get lp_pred_outputs 0) nd2)
+    ;; n is divided by 2 every loop
+    (= nd2 (Get lp_pred_outputs (+ j 1)))
+    ;; In the odd/false branch, we look for an n
+    (= (Get (Arg _ty4 _ctx4) i) (Get oddbr j))
+) (
+    (NTZIterations outerif n j)
+) :ruleset hacker)
+
+;; Identify and optimize lowbit
+
+(rule (
+    (NTZIterations outerif n i)
+    (= outerif (If cond inputs evenbr oddbr))
+    ;; In the even branch, it returns a value that doubles every iter
+    (= evenbr (DoWhile lp_inputs lp_pred_outputs))
+    (= (Const (Int 1) _ty1 _ctx1) (Get lp_inputs j))
+    (= two (Const (Int 2) _ty2 _ctx2))
+    (= (Bop (Mul) (Get (Arg _ty3 _ctx3) j) two) (Get lp_pred_outputs (+ j 1)))    
+    ;; In the odd branch, it returns an 1
+    (= (Const (Int 1) _ty0 _ctx0) (Get oddbr j))
+) (
+    (let lowbitn (Bop (Bitand) n (Uop (Neg) n)))
+    (union (Get outerif j) lowbitn)
+    (union (Get outerif i) (Bop (Div) n lowbitn))
+) :ruleset hacker)
+
+;; Try to do a state-edge-passthrough for loops
+;; NLZIterations guarantees termination for non-zero values
+;; lowbit(0) is undefined behavior
+
+(constructor DummyLoopContext (Expr Expr Expr) Assumption)
+
+(rule (
+    (NTZIterations anyif n i)
+    (= anyif (If cond inputs thenbr elsebr))
+    (= thenbr (DoWhile lpinputs pred_outputs))
+    (= (Get pred_outputs (+ j 1)) (Get (Arg arg_ty then_ctx) j))
+    (HasType (Get pred_outputs (+ j 1)) (Base (StateT)))
+) (
+    (let newlpinputs (TupleRemoveAt lpinputs j))
+    (let newpred_outputs (TupleRemoveAt pred_outputs (+ j 1)))
+    
+    (let newlpctx (DummyLoopContext newlpinputs newpred_outputs pred_outputs))
```

Comment:
These days we've decided to use `TmpCtx` instead, then make sure to `delete` it at the end of the rule. It's safer in case you forget to include some important information in the dummy context